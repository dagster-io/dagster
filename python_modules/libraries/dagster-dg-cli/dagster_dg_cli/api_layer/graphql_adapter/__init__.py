"""GraphQL adapter layer for Dagster Plus CLI."""
