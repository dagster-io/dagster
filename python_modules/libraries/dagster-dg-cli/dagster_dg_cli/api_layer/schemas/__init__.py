"""Pydantic schemas for Dagster Plus API."""
