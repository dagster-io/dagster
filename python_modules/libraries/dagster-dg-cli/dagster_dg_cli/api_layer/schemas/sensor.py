"""Sensor metadata schema definitions."""

from enum import Enum
from typing import Optional

from pydantic import BaseModel


class DgApiSensorStatus(str, Enum):
    """Sensor execution status."""

    RUNNING = "RUNNING"
    STOPPED = "STOPPED"
    PAUSED = "PAUSED"


class DgApiSensorType(str, Enum):
    """Sensor type classification."""

    STANDARD = "STANDARD"
    MULTI_ASSET = "MULTI_ASSET"
    FRESHNESS_POLICY = "FRESHNESS_POLICY"
    AUTO_MATERIALIZE = "AUTO_MATERIALIZE"
    ASSET = "ASSET"


class DgApiSensor(BaseModel):
    """Single sensor metadata model."""

    id: str
    name: str
    status: DgApiSensorStatus
    sensor_type: DgApiSensorType
    description: Optional[str] = None
    repository_origin: Optional[str] = None
    next_tick_timestamp: Optional[float] = None  # Unix timestamp

    class Config:
        from_attributes = True


class DgApiSensorList(BaseModel):
    """GET /api/sensors response."""

    items: list[DgApiSensor]
    total: int
