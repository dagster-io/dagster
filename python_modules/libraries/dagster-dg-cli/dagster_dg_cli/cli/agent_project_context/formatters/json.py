"""JSON formatter for project context."""

import json

from dagster_dg_cli.cli.agent_project_context.formatters import BaseFormatter
from dagster_dg_cli.cli.agent_project_context.models import ProjectContext


class JsonFormatter(BaseFormatter):
    """Formats project context as JSON."""

    def format(self, context: ProjectContext) -> str:
        """Format project context as JSON."""
        # Convert record objects to dictionaries for JSON serialization
        context_dict = {
            "working_directory": context.working_directory,
            "project_root": context.project_root,
            "project_name": context.project_name,
            "components": [
                {
                    "key": comp.key,
                    "summary": comp.summary,
                    "namespace": comp.namespace,
                    "name": comp.name,
                }
                for comp in context.components
            ],
            "integrations": [
                {
                    "name": integ.name,
                    "description": integ.description,
                    "pypi": integ.pypi,
                }
                for integ in context.integrations
            ],
            "project_structure": {
                "has_pyproject_toml": context.project_structure.has_pyproject_toml,
                "has_dbt_project": context.project_structure.has_dbt_project,
                "dbt_project_paths": context.project_structure.dbt_project_paths,
                "python_packages": context.project_structure.python_packages,
            },
            "dbt_schemas": {
                key: {
                    "description": schema.description,
                    "scaffold_params_schema": schema.scaffold_params_schema,
                    "component_schema": schema.component_schema,
                }
                for key, schema in context.dbt_schemas.items()
            },
        }
        return json.dumps(context_dict, indent=2)
