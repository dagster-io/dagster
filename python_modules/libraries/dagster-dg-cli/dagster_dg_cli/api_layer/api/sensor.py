"""Sensor API implementation."""

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional

from dagster_dg_cli.api_layer.graphql_adapter.sensor import (
    get_sensor_via_graphql,
    list_sensors_via_graphql,
)
from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.sensor import DgApiSensor, DgApiSensorList


@dataclass(frozen=True)
class DgApiSensorApi:
    """API for sensor operations."""

    client: IGraphQLClient

    def list_sensors(
        self,
        repository_location_name: Optional[str] = None,
        repository_name: Optional[str] = None,
    ) -> "DgApiSensorList":
        """List all sensors, optionally filtered by repository location and name."""
        return list_sensors_via_graphql(
            self.client,
            repository_location_name=repository_location_name,
            repository_name=repository_name,
        )

    def get_sensor(
        self,
        sensor_name: str,
        repository_location_name: str,
        repository_name: str,
    ) -> "DgApiSensor":
        """Get sensor by name and repository details."""
        return get_sensor_via_graphql(
            self.client,
            sensor_name=sensor_name,
            repository_location_name=repository_location_name,
            repository_name=repository_name,
        )

    def get_sensor_by_name(self, sensor_name: str) -> "DgApiSensor":
        """Get sensor by name, searching across all repositories."""
        from dagster_dg_cli.api_layer.graphql_adapter.sensor import get_sensor_by_name_via_graphql

        return get_sensor_by_name_via_graphql(self.client, sensor_name=sensor_name)
