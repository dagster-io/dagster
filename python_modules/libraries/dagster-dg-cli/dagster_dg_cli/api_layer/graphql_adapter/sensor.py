"""GraphQL implementation for sensor operations."""

from typing import TYPE_CHECKING, Any, Optional

from dagster_dg_cli.utils.plus.gql_client import IGraphQLClient

if TYPE_CHECKING:
    from dagster_dg_cli.api_layer.schemas.sensor import DgApiSensor, DgApiSensorList

LIST_SENSORS_QUERY = """
query ListSensors($repositorySelector: RepositorySelector!) {
    sensorsOrError(repositorySelector: $repositorySelector) {
        __typename
        ... on Sensors {
            results {
                id
                name
                sensorState {
                    status
                }
                sensorType
                description
            }
        }
        ... on RepositoryNotFoundError {
            message
        }
        ... on PythonError {
            message
            stack
        }
    }
}
"""

LIST_REPOSITORIES_QUERY = """
query ListRepositories {
    repositoriesOrError {
        __typename
        ... on RepositoryConnection {
            nodes {
                name
                location {
                    name
                }
                sensors {
                    id
                    name
                    sensorState {
                        status
                    }
                    sensorType
                    description
                }
            }
        }
        ... on PythonError {
            message
            stack
        }
    }
}
"""

GET_SENSOR_QUERY = """
query GetSensor($sensorSelector: SensorSelector!) {
    sensorOrError(sensorSelector: $sensorSelector) {
        __typename
        ... on Sensor {
            id
            name
            sensorState {
                status
            }
            sensorType
            description
        }
        ... on SensorNotFoundError {
            message
        }
        ... on PythonError {
            message
            stack
        }
    }
}
"""


def process_repositories_response(graphql_response: dict[str, Any]) -> "DgApiSensorList":
    """Process GraphQL response from repositories query into DgApiSensorList."""
    from dagster_dg_cli.api_layer.schemas.sensor import (
        DgApiSensor,
        DgApiSensorList,
        DgApiSensorStatus,
        DgApiSensorType,
    )

    repositories_result = graphql_response.get("repositoriesOrError")
    if not repositories_result:
        raise Exception("No repositories data in GraphQL response")

    typename = repositories_result.get("__typename")
    if typename != "RepositoryConnection":
        error_msg = repositories_result.get("message", f"GraphQL error: {typename}")
        raise Exception(error_msg)

    repositories_data = repositories_result.get("nodes", [])

    sensors = []
    for repo in repositories_data:
        repo_name = repo.get("name", "")
        location_name = repo.get("location", {}).get("name", "")
        repository_origin = f"{location_name}@{repo_name}" if location_name and repo_name else None

        for s in repo.get("sensors", []):
            sensor_state = s.get("sensorState", {})
            status = sensor_state.get("status", "STOPPED")

            sensor = DgApiSensor(
                id=s["id"],
                name=s["name"],
                status=DgApiSensorStatus(status),
                sensor_type=DgApiSensorType(s.get("sensorType", "STANDARD")),
                description=s.get("description"),
                repository_origin=repository_origin,
                next_tick_timestamp=None,
            )
            sensors.append(sensor)

    return DgApiSensorList(
        items=sensors,
        total=len(sensors),
    )


def process_sensors_response(graphql_response: dict[str, Any]) -> "DgApiSensorList":
    """Process GraphQL response into DgApiSensorList.

    Args:
        graphql_response: Raw GraphQL response containing "sensorsOrError"

    Returns:
        DgApiSensorList: Processed sensor data
    """
    from dagster_dg_cli.api_layer.schemas.sensor import (
        DgApiSensor,
        DgApiSensorList,
        DgApiSensorStatus,
        DgApiSensorType,
    )

    sensors_result = graphql_response.get("sensorsOrError")
    if not sensors_result:
        raise Exception("No sensors data in GraphQL response")

    typename = sensors_result.get("__typename")
    if typename != "Sensors":
        error_msg = sensors_result.get("message", f"GraphQL error: {typename}")
        raise Exception(error_msg)

    sensors_data = sensors_result.get("results", [])

    sensors = []
    for s in sensors_data:
        sensor_state = s.get("sensorState", {})
        status = sensor_state.get("status", "STOPPED")

        sensor = DgApiSensor(
            id=s["id"],
            name=s["name"],
            status=DgApiSensorStatus(status),
            sensor_type=DgApiSensorType(s.get("sensorType", "STANDARD")),
            description=s.get("description"),
            repository_origin=None,
            next_tick_timestamp=None,
        )
        sensors.append(sensor)

    return DgApiSensorList(
        items=sensors,
        total=len(sensors),
    )


def process_sensor_response(graphql_response: dict[str, Any]) -> "DgApiSensor":
    """Process GraphQL response into DgApiSensor.

    Args:
        graphql_response: Raw GraphQL response containing "sensorOrError"

    Returns:
        DgApiSensor: Processed sensor data
    """
    from dagster_dg_cli.api_layer.schemas.sensor import (
        DgApiSensor,
        DgApiSensorStatus,
        DgApiSensorType,
    )

    sensor_result = graphql_response.get("sensorOrError")
    if not sensor_result:
        raise Exception("No sensor data in GraphQL response")

    typename = sensor_result.get("__typename")
    if typename != "Sensor":
        error_msg = sensor_result.get("message", f"GraphQL error: {typename}")
        raise Exception(error_msg)

    sensor_state = sensor_result.get("sensorState", {})
    status = sensor_state.get("status", "STOPPED")

    return DgApiSensor(
        id=sensor_result["id"],
        name=sensor_result["name"],
        status=DgApiSensorStatus(status),
        sensor_type=DgApiSensorType(sensor_result.get("sensorType", "STANDARD")),
        description=sensor_result.get("description"),
        repository_origin=None,
        next_tick_timestamp=None,
    )


def list_sensors_via_graphql(
    client: IGraphQLClient,
    repository_location_name: Optional[str] = None,
    repository_name: Optional[str] = None,
) -> "DgApiSensorList":
    """Fetch sensors using GraphQL."""
    if repository_location_name and repository_name:
        variables = {
            "repositorySelector": {
                "repositoryLocationName": repository_location_name,
                "repositoryName": repository_name,
            }
        }
        result = client.execute(LIST_SENSORS_QUERY, variables)
        return process_sensors_response(result)
    else:
        result = client.execute(LIST_REPOSITORIES_QUERY)
        return process_repositories_response(result)


def get_sensor_via_graphql(
    client: IGraphQLClient,
    sensor_name: str,
    repository_location_name: str,
    repository_name: str,
) -> "DgApiSensor":
    """Get single sensor via GraphQL."""
    variables = {
        "sensorSelector": {
            "sensorName": sensor_name,
            "repositoryLocationName": repository_location_name,
            "repositoryName": repository_name,
        }
    }

    try:
        result = client.execute(GET_SENSOR_QUERY, variables)
        if (
            result.get("data", {}).get("sensorOrError", {}).get("__typename")
            == "SensorNotFoundError"
        ):
            raise Exception(f"Sensor not found: {sensor_name}")
        return process_sensor_response(result)
    except Exception:
        raise


def get_sensor_by_name_via_graphql(client: IGraphQLClient, sensor_name: str) -> "DgApiSensor":
    """Get sensor by name, searching across all repositories."""
    result = client.execute(LIST_REPOSITORIES_QUERY)
    all_sensors = process_repositories_response(result)

    matching_sensors = [sensor for sensor in all_sensors.items if sensor.name == sensor_name]

    if not matching_sensors:
        raise Exception(f"Sensor not found: {sensor_name}")

    if len(matching_sensors) > 1:
        repo_origins = [sensor.repository_origin or "unknown" for sensor in matching_sensors]
        raise Exception(
            f"Multiple sensors found with name '{sensor_name}' in repositories: {', '.join(repo_origins)}"
        )

    return matching_sensors[0]
