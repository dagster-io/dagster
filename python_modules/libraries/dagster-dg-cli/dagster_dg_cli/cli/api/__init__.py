"""API commands for interacting with Dagster Plus REST-like interface."""

from dagster_dg_cli.cli.api.cli_group import api_group

__all__ = ["api_group"]
