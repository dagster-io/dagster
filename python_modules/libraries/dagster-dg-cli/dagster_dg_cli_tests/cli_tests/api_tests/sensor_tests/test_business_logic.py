"""Test sensor business logic functions without mocks.

These tests focus on testing pure functions that process data without requiring
GraphQL client mocking or external dependencies.
"""

import json

from dagster_dg_cli.api_layer.graphql_adapter.sensor import (
    process_repositories_response,
    process_sensor_response,
    process_sensors_response,
)
from dagster_dg_cli.api_layer.schemas.sensor import (
    DgApiSensor,
    DgApiSensorList,
    DgApiSensorStatus,
    DgApiSensorType,
)
from dagster_dg_cli.cli.api.sensor import format_sensor, format_sensors


class TestProcessSensorResponses:
    """Test the pure functions that process GraphQL responses."""

    def test_process_sensors_response_success(self, snapshot):
        """Test processing a successful sensors GraphQL response."""
        # Sample GraphQL response structure
        response = {
            "sensorsOrError": {
                "__typename": "Sensors",
                "results": [
                    {
                        "id": "sensor1-id",
                        "name": "daily_sensor",
                        "sensorState": {"status": "RUNNING"},
                        "sensorType": "STANDARD",
                        "description": "Daily processing sensor",
                    },
                    {
                        "id": "sensor2-id",
                        "name": "asset_sensor",
                        "sensorState": {"status": "STOPPED"},
                        "sensorType": "ASSET",
                        "description": "Asset change sensor",
                    },
                ],
            }
        }
        result = process_sensors_response(response)

        # Snapshot the entire result to capture structure and data
        snapshot.assert_match(result)

    def test_process_sensors_response_empty(self, snapshot):
        """Test processing an empty sensors GraphQL response."""
        response = {"sensorsOrError": {"__typename": "Sensors", "results": []}}
        result = process_sensors_response(response)

        # Snapshot empty result
        snapshot.assert_match(result)

    def test_process_sensors_response_missing_key(self, snapshot):
        """Test processing a response missing the sensorsOrError key."""
        malformed_response = {}
        try:
            result = process_sensors_response(malformed_response)
            # If no exception, snapshot the result
            snapshot.assert_match(result)
        except Exception as e:
            # Snapshot the error message
            snapshot.assert_match({"error": str(e)})

    def test_process_sensors_response_error_typename(self, snapshot):
        """Test processing a sensors response with error typename."""
        error_response = {
            "sensorsOrError": {
                "__typename": "RepositoryNotFoundError",
                "message": "Repository not found",
            }
        }
        try:
            result = process_sensors_response(error_response)
            snapshot.assert_match(result)
        except Exception as e:
            snapshot.assert_match({"error": str(e)})

    def test_process_repositories_response_success(self, snapshot):
        """Test processing a successful repositories GraphQL response."""
        response = {
            "repositoriesOrError": {
                "__typename": "RepositoryConnection",
                "nodes": [
                    {
                        "name": "main_repo",
                        "location": {"name": "main_location"},
                        "sensors": [
                            {
                                "id": "sensor1-id",
                                "name": "daily_sensor",
                                "sensorState": {"status": "RUNNING"},
                                "sensorType": "STANDARD",
                                "description": "Daily processing sensor",
                            },
                            {
                                "id": "sensor2-id",
                                "name": "asset_sensor",
                                "sensorState": {"status": "PAUSED"},
                                "sensorType": "MULTI_ASSET",
                                "description": "Asset change sensor",
                            },
                        ],
                    },
                    {
                        "name": "secondary_repo",
                        "location": {"name": "secondary_location"},
                        "sensors": [
                            {
                                "id": "sensor3-id",
                                "name": "hourly_sensor",
                                "sensorState": {"status": "STOPPED"},
                                "sensorType": "FRESHNESS_POLICY",
                                "description": None,
                            }
                        ],
                    },
                ],
            }
        }
        result = process_repositories_response(response)

        # Snapshot the entire result to capture structure and data
        snapshot.assert_match(result)

    def test_process_repositories_response_empty(self, snapshot):
        """Test processing an empty repositories GraphQL response."""
        response = {"repositoriesOrError": {"__typename": "RepositoryConnection", "nodes": []}}
        result = process_repositories_response(response)

        snapshot.assert_match(result)

    def test_process_sensor_response_success(self, snapshot):
        """Test processing a successful single sensor GraphQL response."""
        response = {
            "sensorOrError": {
                "__typename": "Sensor",
                "id": "single-sensor-id",
                "name": "critical_sensor",
                "sensorState": {"status": "RUNNING"},
                "sensorType": "AUTO_MATERIALIZE",
                "description": "Critical production sensor",
            }
        }
        result = process_sensor_response(response)

        snapshot.assert_match(result)

    def test_process_sensor_response_not_found_error(self, snapshot):
        """Test processing a sensor not found error."""
        error_response = {
            "sensorOrError": {
                "__typename": "SensorNotFoundError",
                "message": "Sensor 'nonexistent_sensor' not found",
            }
        }
        try:
            result = process_sensor_response(error_response)
            snapshot.assert_match(result)
        except Exception as e:
            snapshot.assert_match({"error": str(e)})


class TestFormatSensors:
    """Test the sensor formatting functions."""

    def _create_sample_sensor_list(self):
        """Create sample DgApiSensorList for testing."""
        sensors = [
            DgApiSensor(
                id="sensor1-id",
                name="daily_sensor",
                status=DgApiSensorStatus.RUNNING,
                sensor_type=DgApiSensorType.STANDARD,
                description="Daily processing sensor",
                repository_origin="main_location@main_repo",
                next_tick_timestamp=1705311000.0,  # 2024-01-15T10:30:00Z
            ),
            DgApiSensor(
                id="sensor2-id",
                name="asset_sensor",
                status=DgApiSensorStatus.STOPPED,
                sensor_type=DgApiSensorType.ASSET,
                description="Asset change sensor",
                repository_origin="main_location@main_repo",
                next_tick_timestamp=None,
            ),
            DgApiSensor(
                id="sensor3-id",
                name="minimal_sensor",
                status=DgApiSensorStatus.PAUSED,
                sensor_type=DgApiSensorType.MULTI_ASSET,
                description=None,
                repository_origin=None,
                next_tick_timestamp=None,
            ),
        ]
        return DgApiSensorList(items=sensors, total=len(sensors))

    def _create_sample_sensor(self):
        """Create sample DgApiSensor for testing."""
        return DgApiSensor(
            id="single-sensor-id",
            name="critical_sensor",
            status=DgApiSensorStatus.RUNNING,
            sensor_type=DgApiSensorType.AUTO_MATERIALIZE,
            description="Critical production sensor",
            repository_origin="prod_location@prod_repo",
            next_tick_timestamp=1705311900.0,  # 2024-01-15T10:45:00Z
        )

    def test_format_sensors_text_output(self, snapshot):
        """Test formatting sensors as text."""
        from dagster_shared.utils.timing import fixed_timezone

        sensor_list = self._create_sample_sensor_list()
        with fixed_timezone("UTC"):
            result = format_sensors(sensor_list, as_json=False)

        # Snapshot the entire text output
        snapshot.assert_match(result)

    def test_format_sensors_json_output(self, snapshot):
        """Test formatting sensors as JSON."""
        sensor_list = self._create_sample_sensor_list()
        result = format_sensors(sensor_list, as_json=True)

        # For JSON, we want to snapshot the parsed structure to avoid formatting differences
        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_sensor_text_output(self, snapshot):
        """Test formatting single sensor as text."""
        from dagster_shared.utils.timing import fixed_timezone

        sensor = self._create_sample_sensor()
        with fixed_timezone("UTC"):
            result = format_sensor(sensor, as_json=False)

        # Snapshot the text output
        snapshot.assert_match(result)

    def test_format_sensor_json_output(self, snapshot):
        """Test formatting single sensor as JSON."""
        sensor = self._create_sample_sensor()
        result = format_sensor(sensor, as_json=True)

        # For JSON, we want to snapshot the parsed structure to avoid formatting differences
        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_format_minimal_sensor_text_output(self, snapshot):
        """Test formatting minimal sensor as text."""
        from dagster_shared.utils.timing import fixed_timezone

        sensor = DgApiSensor(
            id="minimal-id",
            name="minimal_sensor",
            status=DgApiSensorStatus.PAUSED,
            sensor_type=DgApiSensorType.STANDARD,
            description=None,
            repository_origin=None,
            next_tick_timestamp=None,
        )
        with fixed_timezone("UTC"):
            result = format_sensor(sensor, as_json=False)

        snapshot.assert_match(result)

    def test_format_minimal_sensor_json_output(self, snapshot):
        """Test formatting minimal sensor as JSON."""
        sensor = DgApiSensor(
            id="minimal-id",
            name="minimal_sensor",
            status=DgApiSensorStatus.PAUSED,
            sensor_type=DgApiSensorType.STANDARD,
            description=None,
            repository_origin=None,
            next_tick_timestamp=None,
        )
        result = format_sensor(sensor, as_json=True)

        parsed = json.loads(result)
        snapshot.assert_match(parsed)


class TestSensorDataProcessing:
    """Test processing of sensor data structures.

    This class tests pure functions and domain model functionality
    without requiring external dependencies.
    """

    def test_sensor_creation_with_all_fields(self, snapshot):
        """Test creating sensor with all possible fields."""
        sensor = DgApiSensor(
            id="complete-sensor-xyz",
            name="comprehensive_sensor",
            status=DgApiSensorStatus.RUNNING,
            sensor_type=DgApiSensorType.FRESHNESS_POLICY,
            description="Comprehensive test sensor with all fields",
            repository_origin="test_location@test_repo",
            next_tick_timestamp=1705311000.0,
        )

        # Test JSON serialization works correctly
        result = sensor.model_dump_json(indent=2)
        parsed = json.loads(result)
        snapshot.assert_match(parsed)

    def test_sensor_status_enum_values(self):
        """Test that all expected SensorStatus enum values are available."""
        expected_statuses = ["RUNNING", "STOPPED", "PAUSED"]

        actual_statuses = [status.value for status in DgApiSensorStatus]
        assert set(actual_statuses) == set(expected_statuses)

    def test_sensor_type_enum_values(self):
        """Test that all expected SensorType enum values are available."""
        expected_types = [
            "STANDARD",
            "MULTI_ASSET",
            "FRESHNESS_POLICY",
            "AUTO_MATERIALIZE",
            "ASSET",
        ]

        actual_types = [sensor_type.value for sensor_type in DgApiSensorType]
        assert set(actual_types) == set(expected_types)

    def test_sensor_with_missing_optional_fields(self):
        """Test sensor creation with None values for optional fields."""
        sensor = DgApiSensor(
            id="sparse-sensor-123",
            name="sparse_sensor",
            status=DgApiSensorStatus.STOPPED,
            sensor_type=DgApiSensorType.STANDARD,
            description=None,
            repository_origin=None,
            next_tick_timestamp=None,
        )

        assert sensor.id == "sparse-sensor-123"
        assert sensor.name == "sparse_sensor"
        assert sensor.status == DgApiSensorStatus.STOPPED
        assert sensor.sensor_type == DgApiSensorType.STANDARD
        assert sensor.description is None
        assert sensor.repository_origin is None
        assert sensor.next_tick_timestamp is None

    def test_sensor_list_creation(self):
        """Test SensorList creation and basic functionality."""
        sensors = [
            DgApiSensor(
                id="sensor1",
                name="first_sensor",
                status=DgApiSensorStatus.RUNNING,
                sensor_type=DgApiSensorType.STANDARD,
            ),
            DgApiSensor(
                id="sensor2",
                name="second_sensor",
                status=DgApiSensorStatus.PAUSED,
                sensor_type=DgApiSensorType.ASSET,
            ),
        ]
        sensor_list = DgApiSensorList(items=sensors, total=len(sensors))

        assert len(sensor_list.items) == 2
        assert sensor_list.total == 2
        assert sensor_list.items[0].name == "first_sensor"
        assert sensor_list.items[1].name == "second_sensor"

    def test_sensor_timestamp_handling(self, snapshot):
        """Test sensor with various timestamp values."""
        test_cases = [
            # Normal timestamp
            DgApiSensor(
                id="sensor1",
                name="normal_timestamp",
                status=DgApiSensorStatus.RUNNING,
                sensor_type=DgApiSensorType.STANDARD,
                next_tick_timestamp=1705311000.0,
            ),
            # No timestamp
            DgApiSensor(
                id="sensor2",
                name="no_timestamp",
                status=DgApiSensorStatus.STOPPED,
                sensor_type=DgApiSensorType.STANDARD,
                next_tick_timestamp=None,
            ),
            # Future timestamp
            DgApiSensor(
                id="sensor3",
                name="future_timestamp",
                status=DgApiSensorStatus.PAUSED,
                sensor_type=DgApiSensorType.STANDARD,
                next_tick_timestamp=2000000000.0,  # Far future
            ),
        ]

        results = []
        for sensor in test_cases:
            # Test serialization
            serialized = json.loads(sensor.model_dump_json())
            results.append(serialized)

        snapshot.assert_match(results)
