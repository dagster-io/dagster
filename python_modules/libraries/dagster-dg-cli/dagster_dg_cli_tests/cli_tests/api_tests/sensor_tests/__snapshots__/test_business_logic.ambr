# serializer version: 1
# name: TestFormatSensors.test_format_minimal_sensor_json_output[0]
  dict({
    'description': None,
    'name': 'minimal_sensor',
    'next_tick_timestamp': None,
    'sensor_type': 'STANDARD',
    'status': 'PAUSED',
  })
# ---
# name: TestFormatSensors.test_format_minimal_sensor_text_output[0]
  '''
  Name: minimal_sensor
  Status: PAUSED
  Type: STANDARD
  Description: None
  '''
# ---
# name: TestFormatSensors.test_format_sensor_json_output[0]
  dict({
    'description': 'Critical production sensor',
    'name': 'critical_sensor',
    'next_tick_timestamp': 1705311900.0,
    'sensor_type': 'AUTO_MATERIALIZE',
    'status': 'RUNNING',
  })
# ---
# name: TestFormatSensors.test_format_sensor_text_output[0]
  '''
  Name: critical_sensor
  Status: RUNNING
  Type: AUTO_MATERIALIZE
  Description: Critical production sensor
  Next Tick: 2024-01-15 09:45:00
  '''
# ---
# name: TestFormatSensors.test_format_sensors_json_output[0]
  dict({
    'items': list([
      dict({
        'description': 'Daily processing sensor',
        'name': 'daily_sensor',
        'next_tick_timestamp': 1705311000.0,
        'sensor_type': 'STANDARD',
        'status': 'RUNNING',
      }),
      dict({
        'description': 'Asset change sensor',
        'name': 'asset_sensor',
        'next_tick_timestamp': None,
        'sensor_type': 'ASSET',
        'status': 'STOPPED',
      }),
      dict({
        'description': None,
        'name': 'minimal_sensor',
        'next_tick_timestamp': None,
        'sensor_type': 'MULTI_ASSET',
        'status': 'PAUSED',
      }),
    ]),
    'total': 3,
  })
# ---
# name: TestFormatSensors.test_format_sensors_text_output[0]
  '''
  Name: daily_sensor
  Status: RUNNING
  Type: STANDARD
  Description: Daily processing sensor
  Next Tick: 2024-01-15 09:30:00
  
  Name: asset_sensor
  Status: STOPPED
  Type: ASSET
  Description: Asset change sensor
  
  Name: minimal_sensor
  Status: PAUSED
  Type: MULTI_ASSET
  Description: None
  '''
# ---
# name: TestProcessSensorResponses.test_process_repositories_response_empty[0]
  DgApiSensorList(items=[], total=0)
# ---
# name: TestProcessSensorResponses.test_process_repositories_response_success[0]
  DgApiSensorList(items=[DgApiSensor(id='sensor1-id', name='daily_sensor', status=<DgApiSensorStatus.RUNNING: 'RUNNING'>, sensor_type=<DgApiSensorType.STANDARD: 'STANDARD'>, description='Daily processing sensor', repository_origin='main_location@main_repo', next_tick_timestamp=None), DgApiSensor(id='sensor2-id', name='asset_sensor', status=<DgApiSensorStatus.PAUSED: 'PAUSED'>, sensor_type=<DgApiSensorType.MULTI_ASSET: 'MULTI_ASSET'>, description='Asset change sensor', repository_origin='main_location@main_repo', next_tick_timestamp=None), DgApiSensor(id='sensor3-id', name='hourly_sensor', status=<DgApiSensorStatus.STOPPED: 'STOPPED'>, sensor_type=<DgApiSensorType.FRESHNESS_POLICY: 'FRESHNESS_POLICY'>, description=None, repository_origin='secondary_location@secondary_repo', next_tick_timestamp=None)], total=3)
# ---
# name: TestProcessSensorResponses.test_process_sensor_response_not_found_error[0]
  dict({
    'error': "Sensor 'nonexistent_sensor' not found",
  })
# ---
# name: TestProcessSensorResponses.test_process_sensor_response_success[0]
  DgApiSensor(id='single-sensor-id', name='critical_sensor', status=<DgApiSensorStatus.RUNNING: 'RUNNING'>, sensor_type=<DgApiSensorType.AUTO_MATERIALIZE: 'AUTO_MATERIALIZE'>, description='Critical production sensor', repository_origin=None, next_tick_timestamp=None)
# ---
# name: TestProcessSensorResponses.test_process_sensors_response_empty[0]
  DgApiSensorList(items=[], total=0)
# ---
# name: TestProcessSensorResponses.test_process_sensors_response_error_typename[0]
  dict({
    'error': 'Repository not found',
  })
# ---
# name: TestProcessSensorResponses.test_process_sensors_response_missing_key[0]
  dict({
    'error': 'No sensors data in GraphQL response',
  })
# ---
# name: TestProcessSensorResponses.test_process_sensors_response_success[0]
  DgApiSensorList(items=[DgApiSensor(id='sensor1-id', name='daily_sensor', status=<DgApiSensorStatus.RUNNING: 'RUNNING'>, sensor_type=<DgApiSensorType.STANDARD: 'STANDARD'>, description='Daily processing sensor', repository_origin=None, next_tick_timestamp=None), DgApiSensor(id='sensor2-id', name='asset_sensor', status=<DgApiSensorStatus.STOPPED: 'STOPPED'>, sensor_type=<DgApiSensorType.ASSET: 'ASSET'>, description='Asset change sensor', repository_origin=None, next_tick_timestamp=None)], total=2)
# ---
# name: TestSensorDataProcessing.test_sensor_creation_with_all_fields[0]
  dict({
    'description': 'Comprehensive test sensor with all fields',
    'id': 'complete-sensor-xyz',
    'name': 'comprehensive_sensor',
    'next_tick_timestamp': 1705311000.0,
    'repository_origin': 'test_location@test_repo',
    'sensor_type': 'FRESHNESS_POLICY',
    'status': 'RUNNING',
  })
# ---
# name: TestSensorDataProcessing.test_sensor_timestamp_handling[0]
  list([
    dict({
      'description': None,
      'id': 'sensor1',
      'name': 'normal_timestamp',
      'next_tick_timestamp': 1705311000.0,
      'repository_origin': None,
      'sensor_type': 'STANDARD',
      'status': 'RUNNING',
    }),
    dict({
      'description': None,
      'id': 'sensor2',
      'name': 'no_timestamp',
      'next_tick_timestamp': None,
      'repository_origin': None,
      'sensor_type': 'STANDARD',
      'status': 'STOPPED',
    }),
    dict({
      'description': None,
      'id': 'sensor3',
      'name': 'future_timestamp',
      'next_tick_timestamp': 2000000000.0,
      'repository_origin': None,
      'sensor_type': 'STANDARD',
      'status': 'PAUSED',
    }),
  ])
# ---
